File: stdlib/std/time.c
/* std.time — high-resolution monotonic clock.
 *
 * Returns nanoseconds since an unspecified epoch. The epoch is stable
 * for the lifetime of the process but is NOT wall-clock time. Suitable
 * only for measuring intervals; not for persistence or display.
 */

#include <stdint.h>

#ifdef _WIN32

#ifndef WIN32_LEAN_AND_MEAN
#define WIN32_LEAN_AND_MEAN
#endif
#include <windows.h>

/* QueryPerformanceFrequency is fixed for the lifetime of the system
 * per MSDN — cache it on first call to avoid a syscall per tick. */
static LARGE_INTEGER __flang_qpc_freq = {0};

uint64_t __flang_monotonic_ns(void) {
    LARGE_INTEGER now;
    QueryPerformanceCounter(&now);
    if (__flang_qpc_freq.QuadPart == 0) {
        QueryPerformanceFrequency(&__flang_qpc_freq);
    }
    /* Split the multiply to keep the conversion from overflowing on
     * long-running processes: at QPC frequencies in the 1e7 range,
     * a naive `ticks * 1e9` would wrap after ~10 minutes. */
    uint64_t freq = (uint64_t)__flang_qpc_freq.QuadPart;
    uint64_t ticks = (uint64_t)now.QuadPart;
    uint64_t seconds = ticks / freq;
    uint64_t remainder = ticks % freq;
    return seconds * 1000000000ULL + (remainder * 1000000000ULL) / freq;
}

#else /* POSIX */

#include <time.h>

uint64_t __flang_monotonic_ns(void) {
    struct timespec ts;
    /* CLOCK_MONOTONIC is guaranteed not to jump (no NTP adjustments,
     * no daylight-saving). On Linux this is implemented via the vDSO
     * and is effectively a few cycles; on macOS it routes through
     * mach_absolute_time. */
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (uint64_t)ts.tv_sec * 1000000000ULL + (uint64_t)ts.tv_nsec;
}

#endif

File: stdlib/std/time.f
// std.time — high-resolution monotonic clock for measuring intervals.
//
// `monotonic_ns()` returns nanoseconds since an unspecified epoch that
// is stable for the lifetime of the process. The reading IS NOT
// wall-clock time — it doesn't jump under NTP corrections, daylight
// saving, or manual clock changes — and IS NOT persistable across
// process restarts. Use it to time blocks of code, nothing else.
//
// Backed by `QueryPerformanceCounter` on Windows and
// `clock_gettime(CLOCK_MONOTONIC, ...)` on POSIX via time.c.
//
// A first-class `Time` / `Duration` value type, wall-clock readings,
// formatting, and parsing will live elsewhere when needed — this
// module is intentionally narrow.
//
// Typical use:
//
//   const t0 = monotonic_ns()
//   do_work()
//   const elapsed = elapsed_ns(t0)
//   println($"work took {ns_to_millis(elapsed)} ms".as_view())
//
// Or with the lighter `Stopwatch` wrapper:
//
//   let sw = stopwatch()
//   do_work()
//   const ms = sw.elapsed_millis()

#foreign fn __flang_monotonic_ns() u64

// Read the monotonic clock. Returns nanoseconds since the process's
// monotonic epoch (unspecified — only differences are meaningful).
pub fn monotonic_ns() u64 {
    return __flang_monotonic_ns()
}

// Nanoseconds elapsed from the reading `start` until now. Caller is
// responsible for capturing `start` via `monotonic_ns()` first. If
// `start` is in the future (clock-skew across cores in degenerate
// cases) the result is `0` rather than wrapping.
pub fn elapsed_ns(start: u64) u64 {
    const now = __flang_monotonic_ns()
    if now < start { return 0 }
    return now - start
}

// Convert a nanosecond duration to seconds as f64. Convenience for
// formatting — internal math should stay in ns to avoid float drift.
pub fn ns_to_seconds(ns: u64) f64 {
    return (ns as f64) / 1000000000.0
}

pub fn ns_to_millis(ns: u64) f64 {
    return (ns as f64) / 1000000.0
}

pub fn ns_to_micros(ns: u64) f64 {
    return (ns as f64) / 1000.0
}

// Convenience wrapper that captures a start instant and reports elapsed
// time on demand. Cheap to copy (one u64). Reset via `restart()` to
// reuse the same instance across phases.
pub type Stopwatch = struct {
    start_ns: u64
}

// Start a new stopwatch reading the monotonic clock at construction.
pub fn stopwatch() Stopwatch {
    return .{ start_ns = __flang_monotonic_ns() }
}

// Reset to "now". Use between phases when you want a single Stopwatch
// to time multiple back-to-back intervals.
pub fn restart(self: &Stopwatch) {
    self.start_ns = __flang_monotonic_ns()
}

pub fn elapsed_ns(self: Stopwatch) u64 {
    const now = __flang_monotonic_ns()
    if now < self.start_ns { return 0 }
    return now - self.start_ns
}

pub fn elapsed_micros(self: Stopwatch) f64 {
    return ns_to_micros(self.elapsed_ns())
}

pub fn elapsed_millis(self: Stopwatch) f64 {
    return ns_to_millis(self.elapsed_ns())
}

pub fn elapsed_seconds(self: Stopwatch) f64 {
    return ns_to_seconds(self.elapsed_ns())
}
